### PR TITLE
scram implant

### PR DESCRIPTION
**Pull request by Ilya246**
## About the PR
adds an implanter that randomly teleports you up to 100 tiles:
- attempts to teleport you onto a grid
- attempts to not teleport you into solid objects

from testing on actual maps: often teleports you into hallway/maint, sometimes teleports you into departments or secure areas, rarely teleports you into lattice, very rarely teleports you into open space

## Why / Balance
adds another way for syndicates to flee combat, at a risk, since this can teleport you into space, straight into sec, or at your current position
syndicates need more possibilities to do ninja-like stealth (you're valid but you're stealthy)

## Technical details
contains some copypasted code but un-copypasting it is either unfeasible or would require a large overhaul
i do not believe the search for a valid tile will have a considerable performance impact, as it's only done once in a while

## Media

scramb

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added the syndicate scram implanter, which allows you to instantly escape any situation, at a risk.

### CODE REVIEW

**Comment by Nairodian** [comment]
I feel like three charges is a major pain in the ass for security. Tracking down a guy once can be hard enough, having to repeat that three times because every time he gets into trouble he just disappears without any other counter play sounds like actual hell. Then again, it's risky so maybe, mmrh..

**Comment by Ilya246** [comment]
> I feel like three charges is a major pain in the ass for security. Tracking down a guy once can be hard enough, having to repeat that three times because every time he gets into trouble he just disappears without any other counter play sounds like actual hell. Then again, it's risky so maybe, mmrh..

yeah it can teleport you right back where you were, *into open space*, into singulo, into sec, or into whatever dangerous place
it's also likely to simply teleport you into a hallway where you will be spotted

File: Content.Server/Implants/SubdermalImplantSystem.cs [hunk 1]
```diff
@@ -72,6 +86,54 @@ private void OnActivateImplantEvent(EntityUid uid, SubdermalImplantComponent com
...
+            return;
+
+        var xform = Transform(ent);
+        var entityCoords = xform.Coordinates.ToMap(EntityManager);
+
+        // try to find a valid position to teleport to, teleport to whatever works if we can't
+        var targetCoords = new MapCoordinates();
+        for (var i = 0; i < implant.TeleportAttempts; i++)
+        {
+            var distance = implant.TeleportRadius * _random.NextFloat(MathF.Sqrt(1f)); // to get an uniform distribution
```
Comment (deltanedas) [inline Content.Server/Implants/SubdermalImplantSystem.cs:104]: sqrt 1 is 1

File: Content.Server/Implants/Components/ScramImplantComponent.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+/// <summary>
+/// Randomly teleports entity when triggered.
+/// </summary>
+[RegisterComponent]
+public sealed partial class ScramImplantComponent : Component
+{
+    /// <summary>
+    /// Up to how far to teleport the user
+    /// </summary>
+    [DataField("teleportRadius"), ViewVariables(VVAccess.ReadWrite)]
```
Comment (deltanedas) [inline Content.Server/Implants/Components/ScramImplantComponent.cs:15]: no tags

File: Content.Server/Implants/SubdermalImplantSystem.cs [hunk 3]
```diff
@@ -72,6 +86,54 @@ private void OnActivateImplantEvent(EntityUid uid, SubdermalImplantComponent com
...
+            targetCoords = entityCoords.Offset(_random.NextAngle().ToVec() * distance);
+
+            // prefer teleporting to grids
+            if (!_mapManager.TryFindGridAt(targetCoords, out var gridUid, out var grid))
+                continue;
+
+            // the implant user probably does not want to be in your walls
+            var physQuery = GetEntityQuery<PhysicsComponent>();
```
Comment (deltanedas) [inline Content.Server/Implants/SubdermalImplantSystem.cs:112]: store this on system and set in Initialize

File: Content.Server/Implants/SubdermalImplantSystem.cs [hunk 4]
```diff
@@ -72,6 +86,54 @@ private void OnActivateImplantEvent(EntityUid uid, SubdermalImplantComponent com
...
+
+                if (body.BodyType != BodyType.Static ||
+                    !body.Hard ||
+                    (body.CollisionLayer & (int) CollisionGroup.Impassable) == 0)
+                    continue;
+
+                valid = false;
+                break;
+            }
+            if (!valid)
```
Comment (deltanedas) [inline Content.Server/Implants/SubdermalImplantSystem.cs:127]: if (valid) break; instead

**Review by deltanedas — CHANGES_REQUESTED** [review]

File: Content.Server/Implants/SubdermalImplantSystem.cs (hunk 3 above)
Comment (Ilya246) [reply Content.Server/Implants/SubdermalImplantSystem.cs:112]: all other code i saw (many systems) initialises query on the spot

File: Content.Server/Implants/Components/ScramImplantComponent.cs (hunk 2 above)
Comment (Ilya246) [reply Content.Server/Implants/Components/ScramImplantComponent.cs:15]: tag is what, i forgot

File: Content.Server/Implants/SubdermalImplantSystem.cs (hunk 3 above)
Comment (deltanedas) [reply Content.Server/Implants/SubdermalImplantSystem.cs:112]: bad legacy code noob

File: Content.Server/Implants/Components/ScramImplantComponent.cs (hunk 2 above)
Comment (deltanedas) [reply Content.Server/Implants/Components/ScramImplantComponent.cs:15]: [DataField,

**Comment by IProduceWidgets** [comment]
Another great reason to add shotgun tracking implant rounds.

**Comment by IlyaElDunaev** [comment]
This will improve the gameplay. It will be very useful for nuclear operatives.
It would be great to see a demonstration of the implant
https://github.com/space-wizards/space-station-14/assets/154531074/c314f9d0-bf2d-4d17-bf3a-c7f8df03fd34

**Comment by K-Dynamic** [comment]
Have scram implant leave a temporary teleport trail on their feet for validsalad

**Comment by Hmeister-fake** [comment]
> Have scram implant leave a temporary teleport trail on their feet for validsalad

Like the TF2 Teleport trail? That would look sick

**Comment by Ilya246** [comment]
> Have scram implant leave a temporary teleport trail on their feet for validsalad

no need, i think people just questioning where you appeared from is enough

**Comment by Nairodian** [comment]
I think it does need some kind of indication by sound on use, at least, if it doesn't already.

**Comment by Ilya246** [comment]
> I think it does need some kind of indication by sound on use, at least, if it doesn't already.

already does

**Comment by Nairodian** [comment]
> already does

Very well then, seems interesting to me.

**Comment by Emisse** [comment]
I think this might be more interesting as a one time use and cheaper

**Comment by Ilya246** [comment]
> I think this might be more interesting as a one time use and cheaper

it's somewhat highly likely to kill you, in testing it pretty often teleported me into dangerous areas
i could make it have 2 uses but be 6 or 7 TC, probably 6

**Comment by Emisse** [comment]
2 is fine i think i just think 3 is excessive since it means you have to kill the person to capture them, this makes it more likely for sec to not murder someone since theres a higher chance theyll be out of charges

**Comment by Ilya246** [comment]
> 2 is fine i think i just think 3 is excessive since it means you have to kill the person to capture them, this makes it more likely for sec to not murder someone since theres a higher chance theyll be out of charges

2 seems fair to me, either fluke one charge and be forced to use your second one, or be lucky on your first charge and risk while using the second one

**Comment by Ilya246** [comment]
now 2 uses and 6 TC, seems like a fair price to me considering it's a gamble

**Review by Emisse — APPROVED** [review]

**Comment by Ilya246** [comment]
test fail can't possibly be my fault

**Comment by Ilya246** [comment]
verified: test fail is the fault of `f538a2e34a4b62ac406a5871a3c52e51f8f0fad4`, not mine

**Comment by Hmeister-fake** [comment]
I feel like 6 tc is a tad excessive, i mean you can buy no slips, toolbox and 2 smoke grenades for that and easily do the same thing but more reliably

**Comment by Ilya246** [comment]
> I feel like 6 tc is a tad excessive, i mean you can buy no slips, toolbox and 2 smoke grenades for that and easily do the same thing but more reliably

whilist it's potentially suicide, this allows you to instantly escape ANY situation and can even be used while cuffed

**Pull request merged**